File: cpp/Mod_foliate.cpp
#pragma once

function<Int(Int)> foliate_h(Int k) {
    return [k](Int i) { return k + i; };
}

List2Int foliate(ListInt inp);
List2Int foliate_debug(ListInt inp);

class foliate_class {
    
    //VARIABLES        
    ListInt inp;
    ListInt inp_spnew;
    Int cur;
    Int cur_spnew;
    Int lastMax;
    Int lastMax_spnew;
    List2Int res;
    List2Int res_spnew;
    bool debug_spv;
    
    //PREDICATES
    bool p_empty() {
        return empty(inp);
    }

    //OPERATIONS
    void o_init() {                    
        res_spnew = {};
        lastMax_spnew = 0;
        
        res = res_spnew;
        lastMax = lastMax_spnew;       

        if(debug_spv) {
            cout << "o_init" << " ; "  
             << "" << ";"  << "" << ";"  << lastMax << ";"  << res  
            << endl;   
        }

        if(p_empty()) {
            o_end();
        } else {
            o_next();
        }
    }

    void o_next() {                    
        cur_spnew = first(inp);
        inp_spnew = removeFirst(inp);
        res_spnew = append(res, map<Int,Int>(foliate_h(lastMax), firstNNumbers(cur_spnew)));
        lastMax_spnew = reduce<Int>(sp_max, last(res_spnew));
        
        cur = cur_spnew;
        inp = inp_spnew;
        res = res_spnew;
        lastMax = lastMax_spnew;       

        if(debug_spv) {
            cout << "o_next" << " ; "  
             << inp << ";"  << cur << ";"  << lastMax << ";"  << res  
            << endl;   
        }

        if(p_empty()) {
            o_end();
        } else {
            o_next();
        }
    }

    void o_end() {                                  
        if(debug_spv) {
            cout << "o_end" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << ""  
            << endl;   
        }

        return;
    }
    
    public:
    foliate_class(ListInt constr_inp) : inp(constr_inp)   {  }
        
    List2Int execute(bool debug) {           
        debug_spv = debug;
        
        if(debug_spv) {
            cout << "foliate" << ";"  
            << "inp"<< "; " << "cur"<< "; " << "lastMax"<< "; " << "res" 
            << endl;             
            
            cout << "" << "; "  
             << inp << ";"  << "" << ";"  << "" << ";"  << ""  
            << endl;   
        }
        
        o_init();           
        return res;
    }
        
    
};

List2Int foliate(ListInt inp) {
    foliate_class obj(inp);  
    return obj.execute(false);
}


List2Int foliate_debug(ListInt inp) {
    foliate_class obj(inp);  
    return obj.execute(true);
}

File: start/hs/Mod_foliate.hs
module Mod_foliate ( foliate ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types

foliate_h k i = (k + i)

trace' :: String -> a -> a
trace' str x = (trace str x) 
trace' str x = x

data Data_foliate  = Data_foliate  {
  inp :: ListInt,
  cur :: Int,
  lastMax :: Int,
  res :: List2Int
}

--SIGNATURES
foliate :: ListInt -> List2Int
o_init :: Data_foliate -> List2Int
o_next :: Data_foliate -> List2Int
o_end :: Data_foliate -> List2Int
p_empty :: Data_foliate -> Bool

--DEFINITIONS
foliate inp = (o_init (Data_foliate inp undefined undefined undefined))

p_empty (Data_foliate inp cur lastMax res) =
  (empty inp)

o_init (Data_foliate inp cur lastMax res) = (trace' "o_init" flow)
  where
    res' = []
    lastMax' = 0
    inp' = inp
    cur' = cur
    data' = (Data_foliate inp' cur' lastMax' res')
    flow =
      (if' (p_empty data')
        (o_end data')
        (o_next data')
      )

o_next (Data_foliate inp cur lastMax res) = (trace' "o_next" flow)
  where
    cur' = (first inp)
    inp' = (removeFirst inp)
    res' = (append res (map (foliate_h lastMax) (firstNNumbers cur')))
    lastMax' = (reduce sp_max (last res'))
    data' = (Data_foliate inp' cur' lastMax' res')
    flow =
      (if' (p_empty data')
        (o_end data')
        (o_next data')
      )

o_end (Data_foliate inp cur lastMax res) = (trace' "o_end" flow)
  where
    inp' = inp
    cur' = cur
    lastMax' = lastMax
    res' = res
    data' = (Data_foliate inp' cur' lastMax' res')
    flow =
      res'
